sierra_adapter/sierra_reader_new/sierra_reader.py

```diff
@@ -0,0 +1,173 @@
+import json
+import os
+import uuid
+
+import boto3
+
+from client import catalogue_client
+
+
+def get_windows(event):
+    for record in event["Records"]:
+        yield json.loads(record["Sns"]["Message"])
+
+
+def get_sierra_records(client, window):
+    """
+    Fetch the records from Sierra for the given window.
+    """
+    resource_type = os.environ["RESOURCE_TYPE"]
+    sierra_fields = os.environ["SIERRA_FIELDS"]
+
+    # The updatedDate in Sierra is a timestamp supporting second-level
+    # granularity, so we want a query like this:
+    #
+    #       updatedDate=[2022-07-16T23:59:00+00:00,2022-07-20T00:01:01+00:00]
+    #
+    # This only fetches Sierra records which have been updated but not
+    # deleted.
+    print(f"Fetching updated records for {window}")
+    for record in client.get_objects(
+        f"/{resource_type}",
+        params={
+            "fields": sierra_fields,
+            "updatedDate": f'[{window["start"]},{window["end"]}]',
+        },
+    ):
+        yield record
+
+    dates = {window["start"].split("T")[0], window["end"].split("T")[0]}
+
+    # Because filtering by updatedDate doesn't catch deleted records, we
+    # need to make a separate query that filters on deletedDate.
+    #
+    # The deletedDate in Sierra is only a date supporting day-level
+    # granularity.  We want a query like this:
+    #
+    #       deletedDate=2022-07-16
+    #       deletedDate=[2022-07-16,2022-07-20]
+    #
+    print(f"Fetching deleted records for {dates}")
+    deleted_start, _ = window["start"].split("T")
+    deleted_end, _ = window["end"].split("T")
+
+    if deleted_start == deleted_end:
+        deleted_date = deleted_start
+    else:
+        deleted_date = f"[{deleted_start},{deleted_end}]"
+
+    for record in client.get_objects(
+        f"/{resource_type}",
+        params={"fields": sierra_fields, "deletedDate": deleted_date},
+    ):
+        yield record
+
+
+def to_scala_record(record):
+    """
+    Convert a record into the SierraRecord type expected by the Scala apps.
+
+    Note: we may consider changing this at some point and simplifying
+    what the Scala app expects, but for now this matches the messages sent
+    by the existing Sierra reader.
+    """
+    try:
+        return json.dumps(
+            {
+                "id": record["id"],
+                "data": json.dumps(record),
+                "bibIds": [],
+                "modifiedDate": record["deletedDate"] + "T23:59:59Z",
```

Comment:
What is the significance of marking deletions as happening at the end of the day?

Reply:
Arbitrary choice, but maybe feeds into the comment above.

Reply:
Probably worth a comment in the code to make clear it's arbitrary

Reply:
This was a good question, because I'd already forgotten (but remember when I went to write the comment) that this isn't arbitrary after all!